### PR TITLE
NXsas: review minOccurs and group names

### LINKED ISSUE #522: NXsas: review use of minOccurs on various components
For base classes, minOccurs=0 is the default and means the dataset is optional.  For application definitions, the default is minOccurs=1 which means it is required.  (HDF5 imposes maxOccurs=1 since item names must be unique within a group).

add: NXsas [documentation](https://manual.nexusformat.org/classes/applications/NXsas.html#nxsas)

### PR DESCRIPTION
**Pull request by prjemian**
- close #522
- close https://github.com/nexusformat/NIAC/issues/58

### CODE REVIEW

File: applications/NXsas.nxdl.xml [hunk 1]
```diff
@@ -133,40 +130,34 @@
...
-          units="NX_ANGLE">
-        </field>
-        <field name="rotation_angle" type="NX_FLOAT" units="NX_ANGLE"></field>
-        <field name="aequatorial_angle" type="NX_FLOAT"
-          units="NX_ANGLE">
+        <field name="polar_angle" type="NX_FLOAT" units="NX_ANGLE" minOccurs="0" />
+        <field name="azimuthal_angle" type="NX_FLOAT" units="NX_ANGLE" minOccurs="0" />
+        <field name="rotation_angle" type="NX_FLOAT" units="NX_ANGLE" minOccurs="0" />
+        <field name="aequatorial_angle" type="NX_FLOAT" units="NX_ANGLE" minOccurs="0" />
+        <field name="beam_center_x" type="NX_FLOAT" units="NX_LENGTH">
```
Comment (PeterC-DLS) [inline applications/NXsas.nxdl.xml:137]: Should this be optional as the doc for `data` says it's not necessary. Also, shape of that `data` field is column-major - is that correct?

File: applications/NXsas.nxdl.xml [hunk 2]
```diff
@@ -54,37 +54,34 @@
...
       <enumeration>
         <item value="NXsas"/>
       </enumeration>
     </field>
-    <group type="NXinstrument" name="instrument">
-      <group type="NXsource" name="source">
+    <group type="NXinstrument">
+      <group type="NXsource">
         <field name="type">
           <doc>type of radiation source</doc>
```
Comment (prjemian) [inline applications/NXsas.nxdl.xml:69]: Could be a sentence here.  Consistency throughout.

File: applications/NXsas.nxdl.xml [hunk 3]
```diff
@@ -133,40 +130,34 @@
         <field name="y_pixel_size" type="NX_FLOAT" units="NX_LENGTH">
           <doc>Size of a pixel in y direction</doc>
```
Comment (prjemian) [inline applications/NXsas.nxdl.xml:131]: Make descriptions of x & y match

File: applications/NXsas.nxdl.xml (hunk 1 above)
Comment (prjemian) [reply applications/NXsas.nxdl.xml:137]: The doc for `data` should be adjusted.

File: applications/NXsas.nxdl.xml [hunk 4]
```diff
@@ -110,7 +107,7 @@
           </group>
         </group>
       </group>
-      <group type="NXdetector" name="detector">
+      <group type="NXdetector">
         <field name="data" type="NX_NUMBER" signal="1">
```
Comment (prjemian) [inline applications/NXsas.nxdl.xml:111]: drop the `signal="1"` attribute

File: applications/NXsas.nxdl.xml [hunk 5]
```diff
@@ -184,7 +175,7 @@
         <doc>Total integral monitor counts</doc>
       </field>
     </group>
-    <group type="NXdata" name="data">
+    <group type="NXdata">
       <link name="data" target="/NXentry/NXinstrument/NXdetector/data"></link>
```
Comment (prjemian) [inline applications/NXsas.nxdl.xml:179]: change the XML element closure here

File: applications/NXsas.nxdl.xml (hunk 1 above)
Comment (prjemian) [reply applications/NXsas.nxdl.xml:137]: doc for beam center should indicate that this is determined in data reduction

Comment (prjemian) [reply applications/NXsas.nxdl.xml:137]: We'll accept that the ordering directions (the `dimensions`) as presented should remain.  At least until further review from the community.

Comment (prjemian) [reply applications/NXsas.nxdl.xml:137]: It's necessary to specify which is x and y to coordinate with the pixel dimensions, so as to compute angle and then Q.

File: applications/NXsas.nxdl.xml [hunk 6]
```diff
@@ -29,162 +29,166 @@
...
+        <doc>Preset value for time or monitor.</doc>
       </field>
       <field name="integral" type="NX_FLOAT" units="NX_ANY">
-        <doc>Total integral monitor counts</doc>
+        <doc>Total integral monitor counts.</doc>
       </field>
     </group>
-    <group type="NXdata" name="data">
+    <group type="NXdata">
       <link name="data" target="/NXentry/NXinstrument/NXdetector/data"></link>
```
Comment (prjemian) [inline applications/NXsas.nxdl.xml:192]: With removal of `signal="1"` (above), we should indicate the plottable data here using the `signal` attribute of the NXdata group.  Yet, this is not required.

File: applications/NXsas.nxdl.xml [hunk 7]
```diff
@@ -29,162 +29,166 @@
...
+    a monochromatic beam.
   </doc>
   <group type="NXentry">
-    <attribute name="entry">
-      <doc>
-		NeXus convention is to use ``entry1``, ``entry2``, ... 
-		for analysis software to locate each entry
-      </doc>
-    </attribute>
     <field name="title"/>
```
Comment (prjemian) [inline applications/NXsas.nxdl.xml]: While very important to document, `title` is not necessary for data reduction.

File: applications/NXsas.nxdl.xml [hunk 8]
```diff
@@ -29,163 +29,176 @@
...
           </enumeration>
         </field>
       </group>
-      <group type="NXmonochromator" name="monochromator">
+      <group type="NXmonochromator">
         <field name="wavelength" units="NX_WAVELENGTH" type="NX_FLOAT">
-          <doc>The wavelength of the radiation</doc>
+          <doc>The wavelength of the radiation.</doc>
         </field>
         <field name="wavelength_spread" type="NX_FLOAT">
```
Comment (prjemian) [inline applications/NXsas.nxdl.xml:87]: Not *required* for data reduction.

File: applications/NXsas.nxdl.xml [hunk 9]
```diff
@@ -29,163 +29,176 @@
...
     </field>
-    <group type="NXinstrument" name="instrument">
-      <group type="NXsource" name="source">
+    <group type="NXinstrument">
+      <group type="NXsource">
         <field name="type">
-          <doc>type of radiation source</doc>
+          <doc>Type of radiation source.</doc>
         </field>
         <field name="name">
```
Comment (prjemian) [inline applications/NXsas.nxdl.xml:70]: not _required_ for data reduction

File: applications/NXsas.nxdl.xml (hunk 1 above)
Comment (prjemian) [reply applications/NXsas.nxdl.xml]: @PeterC-DLS By your suggestion, it should be possible to replace both the symbol table and the specification of array dimensions by use of the `SIGNAL_indices` attribute, this allowing the use of additional parameters in the `data` array such a time, temperature, electric field, ...

Comment (prjemian) [reply applications/NXsas.nxdl.xml]: After discussion at NIAC2022, it is not necessary to refactor with `NXdata@SIGNAL_indices` (this use is possible, via NXdata).  Will not change.

File: applications/NXsas.nxdl.xml [hunk 10]
```diff
@@ -29,163 +29,176 @@
...
-	<symbol name="nXPixel">
-	  <doc>nXPixel description</doc>
-	</symbol>
-	<symbol name="nYPixel">
-	  <doc>nYPixel description</doc>
-	</symbol>
+      The symbol(s) listed here will be used below to coordinate fields with the same shape.
+    </doc>
+    <symbol name="nXPixel">
+      <doc>nXPixel description</doc>
```
Comment (benajamin) [inline applications/NXsas.nxdl.xml:35]: This should say "number of pixels in the X-direction" or something similar.

Comment (benajamin) [reply applications/NXsas.nxdl.xml:35]: and the same for Y pixels a few lines down.

**Review by benajamin — APPROVED** [review]
LGTM

**Review by PeterC-DLS — APPROVED** [review]
We defer any consideration of dimension ordering in the `data` field to another issue

**Comment by prjemian** [comment]
@PeterC-DLS @benajamin Thanks for the review!

**Pull request merged**